services/download/src/main/java/app/coronawarn/server/services/download/runner/Download.java

```diff
@@ -0,0 +1,147 @@
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.ApplicationArguments;
+import org.springframework.boot.ApplicationRunner;
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+
+
+/**
+ * This runner retrieves diagnosis key batches.
+ */
+@Component
+@Order(1)
+public class Download implements ApplicationRunner {
+
+  private static final Logger logger = LoggerFactory
+      .getLogger(Download.class);
+
+  private final FederationBatchService federationBatchService;
+  private final DiagnosisKeyService diagnosisKeyService;
+  private final DiagnosisKeyBatchDownloader diagnosisKeyBatchDownloader;
+
+  /**
+   * Creates a Download.
+   */
+  Download(FederationBatchService federationBatchService,
+           DiagnosisKeyService diagnosisKeyService, DiagnosisKeyBatchDownloader diagnosisKeyBatchDownloader) {
+    this.federationBatchService = federationBatchService;
+    this.diagnosisKeyService = diagnosisKeyService;
+    this.diagnosisKeyBatchDownloader = diagnosisKeyBatchDownloader;
+  }
+
+  @Override
+  public void run(ApplicationArguments args) {
+    // TODO UTC handling?
+    LocalDate yesterday = LocalDate.now(); //LocalDate.now().minus(Period.ofDays(1));
+    saveFirstBatchTagForDate(yesterday);
+
+    processErrorFederationBatches();
+    
+    processFederationBatches();
+  }
+
+  private void saveFirstBatchTagForDate(LocalDate date) {
+    Optional<DiagnosisKeyBatchContainer> diagnosisKeyBatchContainerOptional =
+        diagnosisKeyBatchDownloader.downloadBatch(date);
+
+    if (diagnosisKeyBatchContainerOptional.isEmpty()) {
+      return;
+    }
+
+    DiagnosisKeyBatchContainer diagnosisKeyBatchContainer = diagnosisKeyBatchContainerOptional.get();
+    federationBatchService.saveFederationBatch(new FederationBatch(diagnosisKeyBatchContainer.getBatchTag(), date));
+  }
+
+  private void processErrorFederationBatches() {
+    // fetch federation batches with error state
+    // store nextbatchtag, store their diagnosis keys, set status to processed
+    // in case of error: set status to error_wont_fix
+
+    List<FederationBatch> federationBatchesWithErrors =
+        federationBatchService.findByStatus(FederationBatchStatus.ERROR);
+    federationBatchesWithErrors.forEach(federationBatch -> {
+      try {
+        Optional<DiagnosisKeyBatchContainer> diagnosisKeyBatchContainerOptional =
+            diagnosisKeyBatchDownloader.downloadBatch(federationBatch.getDate(), federationBatch.getBatchTag());
+
+        if (diagnosisKeyBatchContainerOptional.isEmpty()) {
+          federationBatchService.markFederationBatchWithStatus(federationBatch, FederationBatchStatus.ERROR_WONT_RETRY);
+        }
+
+        // store nextBatchTag
+        DiagnosisKeyBatchContainer diagnosisKeyBatchContainer = diagnosisKeyBatchContainerOptional.get();
```

Comment:
if the optional is empty the get will throw a runtime exception Maybe we can avoid that by using Optional.ifPresentOrElse